File: shell_wrapper/ahe.cc
// Copyright 2025 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include "shell_wrapper/ahe.h"

#include <sys/types.h>

#include <algorithm>
#include <cmath>
#include <cstddef>
#include <cstdint>
#include <memory>
#include <string>
#include <utility>
#include <vector>

#include "absl/numeric/int128.h"
#include "absl/status/status.h"
#include "absl/status/statusor.h"
#include "absl/strings/str_cat.h"
#include "absl/strings/string_view.h"
#include "absl/types/span.h"
#include "include/cxx.h"
#include "shell_encryption/int256.h"
#include "shell_encryption/multi_party/public_key.h"
#include "shell_encryption/multi_party/public_key_share.h"
#include "shell_encryption/multi_party/public_parameter.h"
#include "shell_encryption/multi_party/recovery.h"
#include "shell_encryption/multi_party/secret_key_share.h"
#include "shell_encryption/rns/coefficient_encoder.h"
#include "shell_encryption/rns/crt_interpolation.h"
#include "shell_encryption/rns/rns_error_params.h"
#include "shell_encryption/rns/rns_modulus.h"
#include "shell_encryption/rns/rns_polynomial.h"
#include "shell_encryption/sampler/discrete_gaussian.h"
#include "shell_wrapper/ahe.rs.h"
#include "shell_wrapper/shell_types.h"
#include "shell_wrapper/single_thread_hkdf.h"

using secure_aggregation::MakeFfiStatus;

// Friend classes to call the private constructor of `PublicKey` and
// `SecretKeyShare` directly with a polynomial obtained from a
// `RnsPolynomialWrapper`, potentially obtained by adding polynomials on the
// Rust side.
namespace rlwe {
namespace multi_party {

class SecretKeyShareRawFactory {
 public:
  static SecretKeyShare<secure_aggregation::ModularInt> Create(
      RnsPolynomial<secure_aggregation::ModularInt> key_b,
      absl::Span<const PrimeModulus<secure_aggregation::ModularInt>* const>
          moduli) {
    // `SecretKeyShare` needs a vector instead of a span.
    std::vector<const PrimeModulus<secure_aggregation::ModularInt>*>
        moduli_vector;
    moduli_vector.insert(moduli_vector.begin(), moduli.begin(), moduli.end());
    return SecretKeyShare<secure_aggregation::ModularInt>(
        std::move(key_b), std::move(moduli_vector));
  };
};

class PublicKeyRawFactory {
 public:
  static PublicKey<secure_aggregation::ModularInt> Create(
      const PublicParameter<secure_aggregation::ModularInt>* public_parameter,
      RnsPolynomial<secure_aggregation::ModularInt> key_b) {
    return PublicKey<secure_aggregation::ModularInt>(public_parameter,
                                                     std::move(key_b));
  };
};

}  // namespace multi_party
}  // namespace rlwe

FfiStatus CreateAhePublicParameters(uint64_t log_n, uint64_t t,
                                    const uint64_t* qs, size_t num_qs,
                                    uint64_t error_variance,
                                    double s_base_flood, double s_flood,
                                    rust::Slice<const uint8_t> seed,
                                    AhePublicParameters* out) {
  if (qs == nullptr || out == nullptr) {
    return MakeFfiStatus(absl::InvalidArgumentError(
        secure_aggregation::kNullPointerErrorMessage));
  };

  // Parse and create RNS context.
  auto rns_context_config =
      secure_aggregation::ParseRnsContextConfig(log_n, t, qs, num_qs);
  auto rns_context = secure_aggregation::RnsContext::CreateForBfv(
      rns_context_config.log_n, rns_context_config.qs,
      /*ps=*/{}, rns_context_config.t);
  if (!rns_context.ok()) {
    return MakeFfiStatus(rns_context.status());
  }
  auto rns_context_ptr = std::make_unique<const secure_aggregation::RnsContext>(
      std::move(rns_context.value()));

  // Sample public polynomial.
  auto rlwe_public_parameter = rlwe::multi_party::
      PublicParameter<secure_aggregation::ModularInt>::CreateFromSeed(
          rns_context_ptr.get(), error_variance,
          std::string(ToAbslStringView(seed)), rlwe::PrngType::PRNG_TYPE_HKDF);
  if (!rlwe_public_parameter.ok()) {
    return MakeFfiStatus(rlwe_public_parameter.status());
  }

  // Initialize encoder, error params, and DG sampler.
  auto encoder =
      rlwe::CoefficientEncoder<secure_aggregation::ModularInt>::Create(
          rns_context_ptr.get());
  if (!encoder.ok()) {
    return MakeFfiStatus(encoder.status());
  }
  int log_t = std::floor(std::log2(static_cast<double>(rns_context_config.t)));
  auto error_params =
      rlwe::RnsErrorParams<secure_aggregation::ModularInt>::Create(
          rns_context_config.log_n, rns_context_ptr->MainPrimeModuli(),
          /*aux_moduli=*/{}, log_t, sqrt(error_variance));
  if (!error_params.ok()) {
    return MakeFfiStatus(error_params.status());
  }
  auto dg_sampler_flood =
      rlwe::DiscreteGaussianSampler<secure_aggregation::Integer>::Create(
          s_base_flood);
  if (!dg_sampler_flood.ok()) {
    return MakeFfiStatus(dg_sampler_flood.status());
  }

  out->rlwe_public_parameter = std::move(rlwe_public_parameter.value());
  out->rns_context = std::move(rns_context_ptr);
  out->encoder = std::make_unique<
      const rlwe::CoefficientEncoder<secure_aggregation::ModularInt>>(
      std::move(encoder.value()));
  out->error_params = std::make_unique<
      const rlwe::RnsErrorParams<secure_aggregation::ModularInt>>(
      std::move(error_params.value()));
  out->dg_sampler_flood = std::move(dg_sampler_flood.value());
  out->s_flood = s_flood;

  // All the unique_ptrs are non-null now.
  return MakeFfiStatus();
}

FfiStatus GenerateSecretKeyShare(const AhePublicParameters& params,
                                 SingleThreadHkdfWrapper* prng,
                                 RnsPolynomialWrapper* out) {
  if (prng == nullptr || prng->ptr == nullptr || out == nullptr) {
    return MakeFfiStatus(absl::InvalidArgumentError(
        secure_aggregation::kNullPointerErrorMessage));
  }

  auto secret_key_share =
      rlwe::multi_party::SecretKeyShare<secure_aggregation::ModularInt>::Sample(
          params.rns_context.get(), prng->ptr.get());
  if (!secret_key_share.ok()) {
    return MakeFfiStatus(secret_key_share.status());
  }
  out->ptr = std::make_unique<secure_aggregation::RnsPolynomial>(
      std::move(secret_key_share->Key()));
  return MakeFfiStatus();
}

FfiStatus GeneratePublicKeyShareWrapper(
    const RnsPolynomialWrapper& secret_key_share,
    const AhePublicParameters& params, SingleThreadHkdfWrapper* prng,
    RnsPolynomialWrapper* public_key_share_b,
    RnsPolynomialWrapper* public_key_share_error,
    RnsPolynomialWrapper* wraparound) {
  if (prng == nullptr || public_key_share_b == nullptr ||
      public_key_share_error == nullptr || secret_key_share.ptr == nullptr ||
      prng->ptr == nullptr || public_key_share_b->ptr == nullptr ||
      public_key_share_error->ptr == nullptr) {
    return MakeFfiStatus(absl::InvalidArgumentError(
        secure_aggregation::kNullPointerErrorMessage));
  }
  secure_aggregation::RnsPolynomial* wraparound_ptr =
      wraparound ? wraparound->ptr.get() : nullptr;
  return MakeFfiStatus(
      rlwe::multi_party::PublicKeyShare<secure_aggregation::ModularInt>::
          CreateExplicit(*secret_key_share.ptr,
                         params.rlwe_public_parameter.get(), prng->ptr.get(),
                         public_key_share_b->ptr.get(),
                         public_key_share_error->ptr.get(), wraparound_ptr));
}

FfiStatus AheEncrypt(const uint64_t* input_values, size_t num_input_values,
                     const RnsPolynomialWrapper& public_key_b,
                     const AhePublicParameters& params,
                     SingleThreadHkdfWrapper* prng,
                     RnsPolynomialWrapper* ciphertext_component_b,
                     RnsPolynomialWrapper* ciphertext_component_a,
                     RnsPolynomialWrapper* ciphertext_secret_r,
                     RnsPolynomialWrapper* ciphertext_error_e,
                     RnsPolynomialWrapper* wraparound) {
  if (input_values == nullptr || prng == nullptr ||
      ciphertext_component_a == nullptr || ciphertext_component_b == nullptr ||
      ciphertext_secret_r == nullptr || ciphertext_error_e == nullptr ||
      public_key_b.ptr == nullptr || prng->ptr == nullptr ||
      ciphertext_component_a->ptr == nullptr ||
      ciphertext_component_b->ptr == nullptr ||
      ciphertext_secret_r->ptr == nullptr ||
      ciphertext_error_e->ptr == nullptr) {
    return MakeFfiStatus(absl::InvalidArgumentError(
        secure_aggregation::kNullPointerErrorMessage));
  }

  // Create a `PublicKey` to call `EncryptExplicit` on. This copies the
  // raw polynomial contained in `public_key_b`.
  auto public_key = rlwe::multi_party::PublicKeyRawFactory::Create(
      params.rlwe_public_parameter.get(), *public_key_b.ptr);

  absl::Span<const secure_aggregation::Integer> messages =
      absl::MakeSpan(input_values, num_input_values);
  secure_aggregation::RnsPolynomial* wraparound_ptr =
      wraparound ? wraparound->ptr.get() : nullptr;
  return MakeFfiStatus(public_key.EncryptExplicit(
      messages, params.encoder.get(), params.error_params.get(),
      prng->ptr.get(), ciphertext_component_b->ptr.get(),
      ciphertext_component_a->ptr.get(), ciphertext_secret_r->ptr.get(),
      ciphertext_error_e->ptr.get(), wraparound_ptr));
}

FfiStatus PartialDecrypt(const RnsPolynomialWrapper& ciphertext_component_a,
                         const RnsPolynomialWrapper& secret_key_share,
                         const AhePublicParameters& params,
                         SingleThreadHkdfWrapper* prng,
                         RnsPolynomialWrapper* out,
                         RnsPolynomialWrapper* error_flood,
                         RnsPolynomialWrapper* wraparound) {
  if (prng == nullptr || ciphertext_component_a.ptr == nullptr ||
      secret_key_share.ptr == nullptr || prng->ptr == nullptr ||
      out == nullptr) {
    return MakeFfiStatus(absl::InvalidArgumentError(
        secure_aggregation::kNullPointerErrorMessage));
  }

  // Create a `SecretKeyShare` to call `PartialDecrypt` on. This copies the
  // raw polynomial and the moduli.
  auto sk = rlwe::multi_party::SecretKeyShareRawFactory::Create(
      *secret_key_share.ptr, params.rlwe_public_parameter->Moduli());

  secure_aggregation::RnsPolynomial* error_flood_ptr =
      error_flood ? error_flood->ptr.get() : nullptr;
  secure_aggregation::RnsPolynomial* wraparound_ptr =
      wraparound ? wraparound->ptr.get() : nullptr;

  auto decryption =
      sk.PartialDecrypt(*ciphertext_component_a.ptr, params.s_flood,
                        params.dg_sampler_flood.get(), prng->ptr.get(),
                        error_flood_ptr, wraparound_ptr);
  if (!decryption.ok()) {
    return MakeFfiStatus(decryption.status());
  }

  out->ptr = std::make_unique<secure_aggregation::RnsPolynomial>(
      std::move(decryption.value()));
  return MakeFfiStatus();
}

FfiStatus RecoverMessages(const RnsPolynomialWrapper& sum_partial_decryptions,
                          const RnsPolynomialWrapper& ciphertext_component_b,
                          const AhePublicParameters& params,
                          size_t output_values_length, uint64_t* output_values,
                          size_t* n_written) {
  if (output_values == nullptr || sum_partial_decryptions.ptr == nullptr ||
      ciphertext_component_b.ptr == nullptr || n_written == nullptr) {
    return MakeFfiStatus(absl::InvalidArgumentError(
        secure_aggregation::kNullPointerErrorMessage));
  }

  absl::StatusOr<std::vector<secure_aggregation::Integer>> messages =
      rlwe::multi_party::RecoverMessagesFromSum<secure_aggregation::ModularInt>(
          *sum_partial_decryptions.ptr, *ciphertext_component_b.ptr,
          *params.rlwe_public_parameter, params.encoder.get());

  if (!messages.ok()) {
    return MakeFfiStatus(messages.status());
  }

  // Copy messages from vector to output buffer.
  *n_written = std::min(output_values_length, messages->size());
  std::copy_n(messages->begin(), *n_written, output_values);
  return MakeFfiStatus();
}

FfiStatus AddInPlace(const AhePublicParameters& params,
                     const RnsPolynomialWrapper& in,
                     RnsPolynomialWrapper* out) {
  if (in.ptr == nullptr || out == nullptr || out->ptr == nullptr) {
    return MakeFfiStatus(absl::InvalidArgumentError(
        secure_aggregation::kNullPointerErrorMessage));
  }
  return MakeFfiStatus(
      out->ptr->AddInPlace(*in.ptr, params.rlwe_public_parameter->Moduli()));
}

FfiStatus CreateZeroRnsPolynomialWrapper(const AhePublicParameters& params,
                                         RnsPolynomialWrapper* out) {
  if (out == nullptr) {
    return MakeFfiStatus(absl::InvalidArgumentError(
        secure_aggregation::kNullPointerErrorMessage));
  }

  auto poly = secure_aggregation::RnsPolynomial::CreateZero(
      params.rlwe_public_parameter->LogN(),
      params.rlwe_public_parameter->Moduli(),
      /*is_ntt=*/true);

  if (!poly.ok()) {
    return MakeFfiStatus(poly.status());
  }
  out->ptr = std::make_unique<secure_aggregation::RnsPolynomial>(
      std::move(poly.value()));
  return MakeFfiStatus();
}

FfiStatus PublicKeyComponentA(const AhePublicParameters& params,
                              RnsPolynomialWrapper* out) {
  if (out == nullptr || params.rlwe_public_parameter == nullptr) {
    return MakeFfiStatus(absl::InvalidArgumentError(
        secure_aggregation::kNullPointerErrorMessage));
  }

  out->ptr = std::make_unique<secure_aggregation::RnsPolynomial>(
      params.rlwe_public_parameter->PublicKeyComponentA());
  return MakeFfiStatus();
}

FfiStatus SFlood(const AhePublicParameters& params, double* out) {
  if (out == nullptr) {
    return MakeFfiStatus(absl::InvalidArgumentError(
        secure_aggregation::kNullPointerErrorMessage));
  }
  *out = params.s_flood;
  return MakeFfiStatus();
}

File: shell_wrapper/ahe.rs
// Copyright 2025 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

//! Rust wrapper around the simplified C++ API for multiparty asymmetric
//! Additive Homomorphic Encryption (AHE).

use shell_types::{
    to_cpp_pointer_len_pair, to_cpp_pointer_len_pair_mut, Moduli, RnsContextRef, RnsPolynomial,
};
use single_thread_hkdf::{SeedWrapper, SingleThreadHkdfWrapper};
use status::rust_status_from_cpp;
use std::marker::PhantomData;
use std::mem::MaybeUninit;

#[cxx::bridge]
mod ffi {
    // Struct containing AHE public parameters, accessible from Rust and C++.
    struct AhePublicParameters {
        rlwe_public_parameter: UniquePtr<ConstModularIntPublicParameter>,
        rns_context: UniquePtr<ConstRnsContext>,
        encoder: UniquePtr<ConstModularIntCoefficientEncoder>,
        error_params: UniquePtr<ConstModularIntRnsErrorParams>,
        dg_sampler_flood: UniquePtr<IntegerDiscreteGaussianSampler>,
        s_flood: f64,
    }

    unsafe extern "C++" {
        include!("shell_wrapper/shell_types.h");
        type FfiStatus = status::ffi::FfiStatus;
        type ModuliWrapper = shell_types::ffi::ModuliWrapper;
        type RnsPolynomialWrapper = shell_types::ffi::RnsPolynomialWrapper;

        include!("shell_wrapper/single_thread_hkdf.h");
        type SingleThreadHkdfWrapper = single_thread_hkdf::SingleThreadHkdfWrapper;

        include!("shell_wrapper/ahe_aliases.h");
        #[namespace = "secure_aggregation"]
        type ConstModularIntPublicParameter;
        #[namespace = "secure_aggregation"]
        type ConstModularIntCoefficientEncoder;
        #[namespace = "secure_aggregation"]
        type ConstModularIntRnsErrorParams;
        #[namespace = "secure_aggregation"]
        type IntegerDiscreteGaussianSampler;
        #[namespace = "secure_aggregation"]
        type ConstRnsContext;

        include!("shell_wrapper/shell_aliases.h");
        #[namespace = "secure_aggregation"]
        type RnsContext = shell_types::ffi::RnsContext;
        #[namespace = "secure_aggregation"]
        type ModularInt;

        include!("shell_wrapper/ahe.h");
        pub unsafe fn CreateAhePublicParameters(
            log_n: u64,
            t: u64,
            qs: *const u64,
            num_qs: usize,
            error_variance: u64,
            s_base_flood: f64,
            s_flood: f64,
            seed: &[u8],
            out: *mut AhePublicParameters,
        ) -> FfiStatus;
        pub unsafe fn CreateModuliWrapperFromAheParams(
            params: &AhePublicParameters,
        ) -> ModuliWrapper;
        pub unsafe fn GetPlaintextModulusFromAheParams(params: &AhePublicParameters) -> u64;
        pub unsafe fn GetRnsContextFromAheParams(params: &AhePublicParameters)
            -> *const RnsContext;
        pub unsafe fn GenerateSecretKeyShare(
            params: &AhePublicParameters,
            prng: *mut SingleThreadHkdfWrapper,
            out: *mut RnsPolynomialWrapper,
        ) -> FfiStatus;
        pub unsafe fn GeneratePublicKeyShareWrapper(
            secret_key_share: &RnsPolynomialWrapper,
            params: &AhePublicParameters,
            prng: *mut SingleThreadHkdfWrapper,
            public_key_share_b: *mut RnsPolynomialWrapper,
            public_key_share_error: *mut RnsPolynomialWrapper,
            wraparound: *mut RnsPolynomialWrapper,
        ) -> FfiStatus;
        pub unsafe fn AheEncrypt(
            input_values: *const u64,
            num_input_values: usize,
            public_key_b: &RnsPolynomialWrapper,
            params: &AhePublicParameters,
            prng: *mut SingleThreadHkdfWrapper,
            ciphertext_component_b: *mut RnsPolynomialWrapper,
            ciphertext_component_a: *mut RnsPolynomialWrapper,
            ciphertext_secret_r: *mut RnsPolynomialWrapper,
            ciphertext_error_e: *mut RnsPolynomialWrapper,
            wraparound: *mut RnsPolynomialWrapper,
        ) -> FfiStatus;
        pub unsafe fn PartialDecrypt(
            ciphertext_component_a: &RnsPolynomialWrapper,
            secret_key_share: &RnsPolynomialWrapper,
            params: &AhePublicParameters,
            prng: *mut SingleThreadHkdfWrapper,
            out: *mut RnsPolynomialWrapper,
            error_flood: *mut RnsPolynomialWrapper,
            wraparound: *mut RnsPolynomialWrapper,
        ) -> FfiStatus;
        pub unsafe fn RecoverMessages(
            sum_partial_decryptions: &RnsPolynomialWrapper,
            ciphertext_component_b: &RnsPolynomialWrapper,
            params: &AhePublicParameters,
            num_output_values: usize,
            output_values: *mut u64,
            n_written: *mut usize,
        ) -> FfiStatus;
        pub unsafe fn CreateZeroRnsPolynomialWrapper(
            params: &AhePublicParameters,
            out: *mut RnsPolynomialWrapper,
        ) -> FfiStatus;
        pub unsafe fn PublicKeyComponentA(
            params: &AhePublicParameters,
            out: *mut RnsPolynomialWrapper,
        ) -> FfiStatus;
        pub unsafe fn SFlood(params: &AhePublicParameters, out: *mut f64) -> FfiStatus;
    }
}

/// Re-export bindings. This wrapper holds a status code and
/// unique pointers to AHE parameters, defined in `ahe.h`.
pub use ffi::AhePublicParameters;

/// Creates new public parameters for AHE.
pub fn create_public_parameters(
    log_n: u64,
    t: u64,
    qs: &[u64],
    error_variance: u64,
    s_base_flood: f64,
    s_flood: f64,
    seed: &SeedWrapper,
) -> Result<AhePublicParameters, status::StatusError> {
    let mut out = MaybeUninit::<AhePublicParameters>::zeroed();
    let (in_ptr, in_len) = to_cpp_pointer_len_pair(qs);
    // SAFETY: No lifetime constraints (the new `AhePublicParameters` does
    // not keep any reference to the seed). Only reads the `qs` buffer within a
    // valid range.
    rust_status_from_cpp(unsafe {
        ffi::CreateAhePublicParameters(
            log_n,
            t,
            in_ptr,
            in_len,
            error_variance,
            s_base_flood,
            s_flood,
            seed.as_bytes().into(),
            out.as_mut_ptr(),
        )
    })?;
    // SAFETY: `out` is safely initialized if we get to this point.
    Ok(unsafe { out.assume_init() })
}

/// Returns RNS moduli, containing pointers to the moduli in the public
/// parameters.
pub fn get_moduli<'a>(params: &'a AhePublicParameters) -> Moduli<'a> {
    // SAFETY: `moduli` contains raw pointers to the moduli in `params`, but the bindings
    // don't know that, so we add a lifetime annotation with PhantomData. After
    // that, both `params` and `moduli` live for at least `'a`.
    let moduli = unsafe { ffi::CreateModuliWrapperFromAheParams(params) };
    Moduli { moduli: moduli, phantom: PhantomData }
}

/// Returns the plaintext modulus.
pub fn get_plaintext_modulus(params: &AhePublicParameters) -> u64 {
    unsafe { ffi::GetPlaintextModulusFromAheParams(params) }
}

/// Returns an RnsContextRef, containing a pointer to the RNS context in the
/// public parameters.
pub fn get_rns_context_ref<'a>(params: &'a AhePublicParameters) -> RnsContextRef<'a> {
    // SAFETY: `rns_context` contains a raw pointer to the RNS context in
    // `params`, but the bindings don't know that, so we add a lifetime annotation
    // with PhantomData. After that, both `params` and `rns_context` live for at
    // least `'a`.
    let rns_context = unsafe { ffi::GetRnsContextFromAheParams(params) };
    RnsContextRef { rns_context: rns_context, phantom: PhantomData }
}

/// Generates a secret key share.
pub fn generate_secret_key(
    params: &AhePublicParameters,
    prng: &mut SingleThreadHkdfWrapper,
) -> Result<RnsPolynomial, status::StatusError> {
    let mut out = MaybeUninit::<RnsPolynomial>::zeroed();
    // SAFETY: No lifetime constraints (the new `RnsPolynomial` is obtained by
    // reading data from `params` and using fresh randomness generated by
    // `prng`).
    rust_status_from_cpp(unsafe { ffi::GenerateSecretKeyShare(params, prng, out.as_mut_ptr()) })?;
    // SAFETY: `out` is safely initialized if we get to this point.
    Ok(unsafe { out.assume_init() })
}

/// Generates a public key share in `public_key_share_b`, and stores the error
/// in `public_key_share_error` (for ZK proofs). Returns an absl status code.
pub fn generate_public_key_share(
    secret_key_share: &RnsPolynomial,
    params: &AhePublicParameters,
    prng: &mut SingleThreadHkdfWrapper,
    public_key_share_b: &mut RnsPolynomial,
    public_key_share_error: &mut RnsPolynomial,
    wraparound_option: Option<&mut RnsPolynomial>,
) -> status::Status {
    let wraparound_ptr = match wraparound_option {
        Some(wraparound_ref) => wraparound_ref,
        None => std::ptr::null_mut(),
    };
    // SAFETY: `public_key_share_b` and `public_key_share_error` are non null, and
    // receive valid `RnsPolynomial`s. No lifetime constraints (the values
    // written in the raw pointers are obtained by reading data from `params` and
    // `secret_key_share`, and using fresh randomness generated by `prng`).
    // `wraparound_ptr` is null-checked on the C++ side and only used if available.
    rust_status_from_cpp(unsafe {
        ffi::GeneratePublicKeyShareWrapper(
            secret_key_share,
            params,
            prng,
            public_key_share_b,
            public_key_share_error,
            wraparound_ptr,
        )
    })
}

/// Encodes the input values and encrypts them using the public key
/// `public_key_b`, which is obtained by summing `public_key_shares_b` from all
/// parties. Stores the two components of the
/// ciphertext in `ciphertext_component_b` (a.k.a. ct0) and
/// `ciphertext_component_a` (a.k.a. ct1). Also stores the secret and error for
/// ZK proofs. Returns an absl status code.
pub fn ahe_encrypt(
    input_values: &[u64],
    public_key_b: &RnsPolynomial,
    params: &AhePublicParameters,
    prng: &mut SingleThreadHkdfWrapper,
    ciphertext_component_b: &mut RnsPolynomial,
    ciphertext_component_a: &mut RnsPolynomial,
    ciphertext_secret_r: &mut RnsPolynomial,
    ciphertext_error_e: &mut RnsPolynomial,
    wraparound_option: Option<&mut RnsPolynomial>,
) -> status::Status {
    let (in_ptr, in_len) = to_cpp_pointer_len_pair(input_values);
    let wraparound_ptr = match wraparound_option {
        Some(wraparound_ref) => wraparound_ref,
        None => std::ptr::null_mut(),
    };
    // SAFETY: the 4 ciphertext pointers are non null, and receive valid
    // `RnsPolynomial`s. `AheEncrypt` reads the `input_values` buffer within a valid
    // range. No lifetime constraints (the new `RnsPolynomial`s are obtained by
    // reading data from `params` and `input_Values`, and using fresh randomness
    // generated by `prng`).
    // `wraparound_ptr` is null-checked on the C++ side and only written to if available.
    rust_status_from_cpp(unsafe {
        ffi::AheEncrypt(
            in_ptr,
            in_len,
            public_key_b,
            params,
            prng,
            ciphertext_component_b,
            ciphertext_component_a,
            ciphertext_secret_r,
            ciphertext_error_e,
            wraparound_ptr,
        )
    })
}

/// Computes the partial decryption of a ciphertext component A.
pub fn partial_decrypt(
    ciphertext_component_a: &RnsPolynomial,
    secret_key_share: &RnsPolynomial,
    params: &AhePublicParameters,
    prng: &mut SingleThreadHkdfWrapper,
    error_flood_option: Option<&mut RnsPolynomial>,
    wraparound_option: Option<&mut RnsPolynomial>,
) -> Result<RnsPolynomial, status::StatusError> {
    let mut out = MaybeUninit::<RnsPolynomial>::zeroed();
    let error_flood_ptr = match error_flood_option {
        Some(error_flood_ref) => error_flood_ref,
        None => std::ptr::null_mut(),
    };
    let wraparound_ptr = match wraparound_option {
        Some(wraparound_ref) => wraparound_ref,
        None => std::ptr::null_mut(),
    };
    // SAFETY: No lifetime constraints (the new `RnsPolynomial` is obtained by
    // reading data from `params` and `ciphertext_component_a` and using fresh
    // randomness generated by `prng`).
    // `error_flood_ptr` and `wraparound_ptr` are null-checked on the C++ side and only written to
    // if available.
    rust_status_from_cpp(unsafe {
        ffi::PartialDecrypt(
            ciphertext_component_a,
            secret_key_share,
            params,
            prng,
            out.as_mut_ptr(),
            error_flood_ptr,
            wraparound_ptr,
        )
    })?;
    // SAFETY: `out` is safely initialized if we get to this point.
    Ok(unsafe { out.assume_init() })
}

/// Recovers messages from a sum of partial decryptions.
pub fn recover_messages(
    sum_partial_decryptions: &RnsPolynomial,
    ciphertext_component_b: &RnsPolynomial,
    params: &AhePublicParameters,
    output_values: &mut [u64],
) -> Result<usize, status::StatusError> {
    let mut n_written = 0usize;
    let (out_ptr, out_len) = to_cpp_pointer_len_pair_mut(output_values);
    // SAFETY:  No lifetime constraints (`BufferResult` just holds two ints and
    // does not keep any reference to the inputs). `Decrypt` only modifies the
    // `output_values` buffer within a valid range.
    rust_status_from_cpp(unsafe {
        ffi::RecoverMessages(
            sum_partial_decryptions,
            ciphertext_component_b,
            params,
            out_len,
            out_ptr,
            &mut n_written,
        )
    })?;
    Ok(n_written)
}

/// Creates a zero polynomial with the same RNS parameters as `params`.
pub fn create_zero_rns_polynomial(
    params: &AhePublicParameters,
) -> Result<RnsPolynomial, status::StatusError> {
    let mut out = MaybeUninit::<RnsPolynomial>::zeroed();
    // SAFETY: No lifetime constraints (the new `RnsPolynomial` is obtained by
    // reading data from `params`).
    rust_status_from_cpp(unsafe { ffi::CreateZeroRnsPolynomialWrapper(params, out.as_mut_ptr()) })?;
    // SAFETY: `out` is safely initialized if we get to this point.
    Ok(unsafe { out.assume_init() })
}

/// Returns the public key component A from the given parameters.
pub fn public_key_component_a(
    params: &AhePublicParameters,
) -> Result<RnsPolynomial, status::StatusError> {
    let mut out = MaybeUninit::<RnsPolynomial>::zeroed();
    // SAFETY: No lifetime constraints. The number of elements written is checked
    // inside the C++ function.
    rust_status_from_cpp(unsafe { ffi::PublicKeyComponentA(params, out.as_mut_ptr()) })?;
    // SAFETY: `out` is safely initialized if we get to this point.
    Ok(unsafe { out.assume_init() })
}

pub fn s_flood(params: &AhePublicParameters) -> Result<f64, status::StatusError> {
    let mut out: f64 = 0.0;
    // SAFETY: No lifetime constraints. Only a member of type double of params is accessed on the
    // C++ side and written to `out`, and no pointers are retained on the C++ side.
    let status = unsafe { ffi::SFlood(params, &mut out) };
    rust_status_from_cpp(status)?;
    return Ok(out);
}
